walletunlocker/service.go

```diff
@@ -408,18 +468,32 @@ func (u *UnlockerService) ChangePassword(ctx context.Context,
 	if err != nil {
 		return nil, err
 	}
-	// Unload the wallet to allow lnd to open it later on.
-	defer loader.UnloadWallet()
-
-	// Since the macaroon database is also encrypted with the wallet's
-	// password, we'll remove all of the macaroon files so that they're
-	// re-generated at startup using the new password. We'll make sure to do
-	// this after unlocking the wallet to ensure macaroon files don't get
-	// deleted with incorrect password attempts.
-	for _, file := range u.macaroonFiles {
-		err := os.Remove(file)
-		if err != nil && !os.IsNotExist(err) {
-			return nil, err
+
+	// Now that we've opened the wallet, we need to close it in case of an
+	// error. But not if we succeed, then the caller must close it.
+	orderlyReturn := false
+	defer func() {
+		if !orderlyReturn {
+			_ = loader.UnloadWallet()
+		}
+	}()
+
+	// Before we actually change the password, we need to check if all flags
+	// were set correctly. The content of the previously generated macaroon
+	// files will become invalid after we generate a new root key. So we try
+	// to delete them here and they will be recreated during normal startup
+	// later. If they are missing, this is only an error if the
+	// stateless_init flag was not set.
+	if in.NewMacaroonRootKey || in.StatelessInit {
```

Comment:
Don't we _always_ need to do this for any change password request?

Reply:
No, not anymore. Before, we just deleted the whole macaroon DB, effectively invalidating all macaroons. Therefore it did make sense to delete the macaroon files themselves as well. But now, we only re-generate the root key if specifically requested. So we only need to remove the macaroon files on disk if the root key is changed or if the user tries to switch to stateless.